java/test_flatbuf1/Test.java

```diff
@@ -16,16 +17,246 @@
     Test() {
         /* Create an instance of the DDlog program with one worker thread. */
         this.api = new DDlogAPI(1, null, false);
-        api.record_commands("replay.dat", false);
+        api.recordCommands("replay.dat", false);
     }
 
     void check(int exit) {
         if (exit != 0)
             throw new RuntimeException("Exit code shows error" + exit);
     }
 
-    void onCommit(DDlogCommand command) {
-        System.out.println(command.toString());
+    
+    // typedef tuple = (bool, bit<8>, string)
+    private String printTuple(Tuple3__bool__bit_8___stringReader t) {
+        return "(" + t.a0() + ", " + t.a1() + ", \"" + t.a2() + "\")";
+    }
+ 
+    // typedef Many = A{x: string}
+    //              | B{b: bool}
+    //              | D{t: tuple}
+    private String printMany(ManyReader m) {
+        if (m instanceof ManyReader.A) {
+            return "A{\"" + ((ManyReader.A)m).x() + "\"}";
+        } else if (m instanceof ManyReader.B) {
+            return "B{" + ((ManyReader.B)m).b() + "}";
+        } else if (m instanceof ManyReader.D) {
+            return "D{" + printTuple(((ManyReader.D)m).t()) + "}";
+        } else {
+            throw new IllegalArgumentException("Invalid Many value " + m.toString());
+        }
+    }
+
+    void onCommit(DDlogCommand<Object> command) {
+        int relid = command.relid();
+        switch (relid) {
+            // output relation BO(b: bool)
+            case flatbufTestRelation.BO: {
+                BOReader v = (BOReader)command.value();
```

Comment:
Alternatively, maybe (if type erasure allows) you could define a method in DDlogCommand: 
`<T> T as() { return (DDlogCommand<T>)this; }`
and then write `command.as<BOReader>().value();`

Reply:
Quick test seems to show that this works. But what's the advantage of this?  It does not seem to be any safer than just directly casting the record type.